### PR TITLE
xc7: Use the common_slice definition when LUTRAM used as a LUT.

### PR DESCRIPTION
**Pull request by mithro**
Reduce duplication by using the
[`xc7/primitives/common_slice/Nlut/ntemplate.Nlut.pb_type.xml`](xc7/primitives/common_slice/Nlut/ntemplate.Nlut.pb_type.xml)
definition when
[`xc7/primitives/slicem/Ndram/ntemplate.Nd_ram.pb_type.xml`](xc7/primitives/slicem/Ndram/ntemplate.Nd_ram.pb_type.xml)
is operating in `LUT` mode.

Signed-off-by: Tim 'mithro' Ansell <me@mith.ro>

### CODE REVIEW

File: .github/kokoro/ice40.sh [hunk 1]
```diff
@@ -11,7 +11,9 @@ echo "Running ice40 tests"
 echo "----------------------------------------"
 (
 	cd build
-	VPR_NUM_WORKERS=${CORES} make -j ${MAX_CORES} --output-sync=target \
-		all_ice40
+	export VPR_NUM_WORKERS=${CORES}
```
Comment (litghost) [inline .github/kokoro/ice40.sh]: This doesn't belong in this PR

File: common/cmake/file_targets.cmake [hunk 2]
```diff
@@ -223,7 +224,7 @@ function(ADD_FILE_TARGET)
   #
   # GENERATED must be passed if the source file is generated and is placed
   # within the binary directory.
-  set(options GENERATED)
+  set(options GENERATED IGNORE_INCLUDES)
```
Comment (litghost) [inline common/cmake/file_targets.cmake]: Why would IGNORE_INCLUDES ever be valid?

File: common/cmake/gen.cmake [hunk 3]
```diff
@@ -182,8 +182,7 @@ function(N_TEMPLATE)
   #   NAME <name>
   #   SRCS <list of sources>
   #   PREFIXES <list of prefixes>
-  #   [APPLY_V2X]
-  #   [APPLY_VERILOG_IMAGE_GEN]
+  #   [GENERATED | SCANNER_TYPE <verilog|xml>]
```
Comment (litghost) [inline common/cmake/gen.cmake]: This is a mistake.  The way it was written was correct and should be left alone.

File: .github/kokoro/ice40.sh (hunk 1 above)
Comment (mithro) [reply .github/kokoro/ice40.sh]: Its in a separate pull request at https://github.com/SymbiFlow/symbiflow-arch-defs/pull/785 that this is based on.

File: common/cmake/gen.cmake (hunk 3 above)
Comment (mithro) [reply common/cmake/gen.cmake]: The way it was written before did not work. If a `ntemplate` file was dependent on another `ntemplate` file the cmake system would just error.

File: common/cmake/file_targets.cmake (hunk 2 above)
Comment (mithro) [reply common/cmake/file_targets.cmake]: The ignores are not valid until after ntemplate substitution has occurred. We **still** need a way to get which dependencies exist so we can perform ntemplate substitution on them as the current system doesn't let us rescan the dependencies after the ntemplate run has occurred.

**Comment by mithro** [comment]
@litghost - FYI This causes `genfasm` to segfault in `handle_fasm_prefix` code;
```
#0  0x0000555555777a51 in fasm::handle_fasm_prefix(t_metadata_dict const*, t_pb_graph_node const*, t_pb_type const*) [clone .isra.194] ()
#1  0x0000555555777e04 in fasm::FasmWriterVisitor::build_clb_prefix[abi:cxx11](t_pb const*, t_pb_graph_node const*) const ()
#2  0x000055555577a976 in fasm::FasmWriterVisitor::visit_all_impl(vtr::flat_map2<int, t_pb_route, std::less<int> > const&, t_pb const*) ()
#3  0x00005555555c1125 in NetlistWalker::walk_blocks(vtr::flat_map2<int, t_pb_route, std::less<int> > const&, t_pb const*) ()
#4  0x00005555555c10b6 in NetlistWalker::walk_blocks(vtr::flat_map2<int, t_pb_route, std::less<int> > const&, t_pb const*) ()
#5  0x00005555555c10b6 in NetlistWalker::walk_blocks(vtr::flat_map2<int, t_pb_route, std::less<int> > const&, t_pb const*) ()
#6  0x00005555555c10b6 in NetlistWalker::walk_blocks(vtr::flat_map2<int, t_pb_route, std::less<int> > const&, t_pb const*) ()
#7  0x00005555555c3c16 in NetlistWalker::walk() ()
#8  0x00005555555b059a in main ()
```

Still investigating why at the moment.

**Comment by mithro** [comment]
With debug now;
```
#0  0x00007ffff6e0859b in std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000555555a48f85 in t_metadata_value::as_string[abi:cxx11]() const (this=0x0) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/libs/libarchfpga/src/physical_types.h:87
#2  0x0000555555a44cf5 in fasm::handle_fasm_prefix (meta=0x55555610aa98, pb_graph_node=0x555556248500, pb_type=0x555556104e18) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/fasm.cpp:119
#3  0x0000555555a4521e in fasm::FasmWriterVisitor::build_clb_prefix[abi:cxx11](t_pb const*, t_pb_graph_node const*) const (this=0x7fffffffaa60, pb=0x5555564680c8, pb_graph_node=0x555556248500) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/fasm.cpp:174
#4  0x0000555555a4570f in fasm::FasmWriterVisitor::visit_all_impl (this=0x7fffffffaa60, pb_routes=..., pb=0x5555564680c8) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/fasm.cpp:223
#5  0x00005555559e7435 in visit_all (this=0x7fffffffaa60, top_pb_route=..., pb=0x5555564680c8) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.h:45
#6  0x00005555559e761c in _ZN13NetlistWalker11walk_blocksERKN3vtr9flat_map2Ii10t_pb_routeSt4lessIiEEEPK4t_pb (this=0x7fffffffaa58, top_pb_route=..., pb=0x5555564680c8) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.cpp:32
#7  0x00005555559e772f in _ZN13NetlistWalker11walk_blocksERKN3vtr9flat_map2Ii10t_pb_routeSt4lessIiEEEPK4t_pb (this=0x7fffffffaa58, top_pb_route=..., pb=0x555556480f88) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.cpp:49
#8  0x00005555559e772f in _ZN13NetlistWalker11walk_blocksERKN3vtr9flat_map2Ii10t_pb_routeSt4lessIiEEEPK4t_pb (this=0x7fffffffaa58, top_pb_route=..., pb=0x55555646b0e8) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.cpp:49
#9  0x00005555559e772f in _ZN13NetlistWalker11walk_blocksERKN3vtr9flat_map2Ii10t_pb_routeSt4lessIiEEEPK4t_pb (this=0x7fffffffaa58, top_pb_route=..., pb=0x55555647cac0) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.cpp:49
#10 0x00005555559e7569 in _ZN13NetlistWalker4walkEv (this=0x7fffffffaa58) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/vpr/src/base/netlist_walker.cpp:19
#11 0x00005555556e152b in write_fasm () at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/main.cpp:45
#12 0x00005555556e16c2 in main (argc=33, argv=0x7fffffffd6c8) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/main.cpp:81
```

**Comment by mithro** [comment]
```
>>> up
#1  0x0000555555a48f85 in t_metadata_value::as_string[abi:cxx11]() const (this=0x0) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/libs/libarchfpga/src/physical_types.h:87
87          std::string as_string() const { return value_; }
>>> 
#2  0x0000555555a44cf5 in fasm::handle_fasm_prefix (meta=0x55555610aa98, pb_graph_node=0x555556248500, pb_type=0x555556104e18) at /usr/local/google/home/tansell/github/verilog-to-routing/vtr-verilog-to-routing/utils/fasm/src/fasm.cpp:119
119       auto fasm_prefix_unsplit = pb_type->meta.one("fasm_prefix")->as_string();
>>> p pb_type
$1 = (const t_pb_type *) 0x555556104e18
>>> p *pb_type
$2 = {
  name = 0x55555610a010 "C_DRAM", 
  num_pb = 1, 
  blif_model = 0x0, 
  model = 0x0, 
  class_type = UNKNOWN_CLASS, 
  modes = 0x55555610a968, 
  num_modes = 3, 
  ports = 0x55555610a290, 
  num_ports = 11, 
  num_clock_pins = 1, 
  num_input_pins = 17, 
  num_output_pins = 5, 
  parent_mode = 0x5555560ed700, 
  depth = 3, 
  annotations = 0x0, 
  num_annotations = 0, 
  pb_type_power = 0x5555561055d0, 
  meta = {
    <std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<t_metadata_value, std::allocator<t_metadata_value> >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::vector<t_metadata_value, std::allocator<t_metadata_value> > > > >> = std::unordered_map with 2 elements = {
      ["subtype"] = std::vector of length 1, capacity 1 = {[0] = {
          value_ = "ignore"
        }},
      ["type"] = std::vector of length 1, capacity 1 = {[0] = {
          value_ = "block"
        }}
    }, <No data fields>}
}
>>> 
```

Looks like `genfasm` is requesting the `fasm_prefix` on a `pb_type` that doesn't have one. Will send a fix shortly.

File: common/cmake/gen.cmake (hunk 3 above)
Comment (litghost) [reply common/cmake/gen.cmake]: The solution to this issue was added by @kgugala in #594, you don't need invent another one.

File: common/cmake/file_targets.cmake (hunk 2 above)
Comment (litghost) [reply common/cmake/file_targets.cmake]: > The ignores are not valid until after ntemplate substitution has occurred.

But this doesn't have to be true, and Karol's was such that the includes were valid before and after.  That avoided this additional complexity, etc etc.

**Comment by mithro** [comment]
I believe this is segfaulting because of https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/62

**Comment by mithro** [comment]
@litghost - I removed the cmake changes, but still needed a change to the ntemplate python script to make this work. I have also needed to updated vpr to fix the segfault in genfasm.

File: common/cmake/file_targets.cmake (hunk 2 above)
Comment (mithro) [reply common/cmake/file_targets.cmake]: Removed these changes for now but the ignore include changes where needed for both this and changes in how muxgen stuff worked in https://github.com/SymbiFlow/symbiflow-arch-defs/pull/703, so it will likely back soon.

**Comment by mithro** [comment]
```
Traceback (most recent call last):
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 126, in <module>
    main()
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 122, in main
    debug=args.debug)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 76, in run
    assembler.parse_fasm_filename(filename_in, extra_features=extra_features)
  File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/prjxray/fasm_assembler.py", line 173, in parse_fasm_filename
    raise FasmLookupError('\n'.join(missing_features))
prjxray.fasm_assembler.FasmLookupError: Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.RAM not found from line 'CLBLM_L_X20Y119.SLICEM_X0.RAM'
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.SMALL not found from line 'CLBLM_L_X20Y119.SLICEM_X0.SMALL'
```

**Comment by mithro** [comment]
I think this error is caused by `PACKAGE_SPEC "vtr 7.0.5_7336_g39fdec17e 20190529_000456"`

**Comment by litghost** [comment]
> ```
> Traceback (most recent call last):
>   File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 126, in <module>
>     main()
>   File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 122, in main
>     debug=args.debug)
>   File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/utils/fasm2frames.py", line 76, in run
>     assembler.parse_fasm_filename(filename_in, extra_features=extra_features)
>   File "/tmpfs/src/github/symbiflow-arch-defs-presubmit-xc7/third_party/prjxray/prjxray/fasm_assembler.py", line 173, in parse_fasm_filename
>     raise FasmLookupError('\n'.join(missing_features))
> prjxray.fasm_assembler.FasmLookupError: Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.RAM not found from line 'CLBLM_L_X20Y119.SLICEM_X0.RAM'
> Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.SMALL not found from line 'CLBLM_L_X20Y119.SLICEM_X0.SMALL'
> ```

This is a result of prefixing being lost.  Let me look at pb_type to see if I can see the issue.

**Comment by litghost** [comment]
The pb_type seems okay, but there may still be a lingering issue with how the prefixes are resolved.  The XML looks okay, so I would suspect a genfasm bug.

**Comment by mithro** [comment]
I'm pretty sure it is caused by https://github.com/SymbiFlow/vtr-verilog-to-routing/commit/b949ac473dedf009c584c2bd1a0ca6d4e13632fc

**Comment by litghost** [comment]
> I'm pretty sure it is caused by [SymbiFlow/vtr-verilog-to-routing@b949ac4](https://github.com/SymbiFlow/vtr-verilog-to-routing/commit/b949ac473dedf009c584c2bd1a0ca6d4e13632fc)

No, it's an old bug I thought I had fixed.  You fixed a latent bug in my change that was aiming to fix it in the first place. 

See: https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/33

**Comment by mithro** [comment]
@litghost Could you look at fixing fasm fix?

**Comment by litghost** [comment]
> @litghost Could you look at fixing fasm fix?

I don't think so, between the look ahead map and routing timing export.

**Comment by litghost** [comment]
> @litghost Could you look at fixing fasm fix?

In the short term, can you leave the fasm_prefix on the pb_type (rather than the mode)?  It will require a change to the LUT ntemplate, but it should be possible.

**Comment by mithro** [comment]
The issue seems to be the following;
```
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.DLUT.INIT[31:0]=32'b10010110101001011001011010100101
CLBLM_L_X20Y119.SLICEM_X0.RAM
CLBLM_L_X20Y119.SLICEM_X0.SMALL
CLBLM_L_X20Y119.SLICEM_X0.CLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.CLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.CLUT.INIT[31:0]=32'b10010110101001011001011010100101
CLBLM_L_X20Y119.SLICEM_X0.BLUT.INIT[31:0]=32'b11111111111111110000000000000000
CLBLM_L_X20Y119.SLICEM_X0.BLUT.INIT[63:32]=32'b00000000000000010000000000000001
CLBLM_L_X20Y119.SLICEM_X0.ALUT.INIT[31:0]=32'b00001111000011110000111100001111
CLBLM_L_X20Y119.SLICEM_X0.ALUT.INIT[63:32]=32'b00000000000000001111111111111111
```

**Comment by litghost** [comment]
Most of it is fine, but this:
```
CLBLM_L_X20Y119.SLICEM_X0.RAM
CLBLM_L_X20Y119.SLICEM_X0.SMALL
``` 

Is missing a prefix, e.g. [ABCD]LUT.

**Comment by mithro** [comment]
I made it print out when `vist_all_impl` sets `clb_prefix_` and got the following output;
```
clb: DLUT.

      RAM
      SMALL
    
clb: DLUT.

      RAM
      SMALL
    
INIT[31:0]=32'b10010110101001011001011010100101

clb: CLUT.
clb: 

      RAM
      SMALL
    
clb: CLUT.

      RAM
      SMALL
```

I'm unclear why I see `clb: DLUT.` twice (which is causing the duplicate DLUT lines). The `clb:` line is the one causing the invalid values.

**Comment by litghost** [comment]
Duplicates are fine, and the duplicate is coming from the WCLK fasm_mux declaration.

**Comment by mithro** [comment]
So, I changed the `<pb_type>` to have the following;
```xml
<pb_type name="{N}_DRAM" num_pb="1" xmlns:xi="http://www.w3.org/2001/XInclude">
  <clock  name="CLK"    num_pins="1" />
  <input  name="A"      num_pins="6" />
  <input  name="WA"     num_pins="8" />
  <input  name="DI1"    num_pins="1" />
  <input  name="DI2"    num_pins="1" />
  <input  name="WE"     num_pins="1" />

  <output name="DO6"    num_pins="1" />
  <output name="DO6_32" num_pins="1" />
  <output name="DO5_32" num_pins="1" />
  <output name="O6"     num_pins="1" />
  <output name="O5"     num_pins="1" />

  <!-- TODO: Missing modes: SRL -->
  <mode name="LUT">
    <xi:include href="../../common_slice/Nlut/ntemplate.Nlut.pb_type.xml" />
    <interconnect>
      <direct name="A1" input="{N}_DRAM.A[0]" output="{N}LUT.A1"/>
      <direct name="A2" input="{N}_DRAM.A[1]" output="{N}LUT.A2"/>
      <direct name="A3" input="{N}_DRAM.A[2]" output="{N}LUT.A3"/>
      <direct name="A4" input="{N}_DRAM.A[3]" output="{N}LUT.A4"/>
      <direct name="A5" input="{N}_DRAM.A[4]" output="{N}LUT.A5"/>
      <direct name="A6" input="{N}_DRAM.A[5]" output="{N}LUT.A6"/>

      <direct name="O5" input="{N}LUT.O5" output="{N}_DRAM.O5"/>
      <direct name="O6" input="{N}LUT.O6" output="{N}_DRAM.O6"/>
    </interconnect>
    <metadata>
      <meta name="fasm_prefix">d{N}LUT</meta>
    </metadata>
  </mode>
  <mode name="64_DUAL_PORT">
    <xi:include href="dpram64.pb_type.xml" />
    <interconnect>
      <direct name="CLK"   input="{N}_DRAM.CLK"     output="DPRAM64.CLK"        />
      <direct name="A"     input="{N}_DRAM.A"       output="DPRAM64.A"          />
      <direct name="WA"    input="{N}_DRAM.WA[5:0]" output="DPRAM64.WA[5:0]"    />
      <direct name="WA7"   input="{N}_DRAM.WA[6]"   output="DPRAM64.WA7"        />
      <direct name="WA8"   input="{N}_DRAM.WA[7]"   output="DPRAM64.WA8"        />
      <direct name="DI"    input="{N}_DRAM.DI1"     output="DPRAM64.DI"         />
      <direct name="WE"    input="{N}_DRAM.WE"      output="DPRAM64.WE"         />

      <direct name="O6"    input="DPRAM64.O"        output="{N}_DRAM.O6"        />
      <direct name="DO6"   input="DPRAM64.O"        output="{N}_DRAM.DO6"       />
    </interconnect>
    <metadata>
      <meta name="fasm_prefix">c{N}LUT</meta>
    </metadata>
  </mode>
  <mode name="32_DUAL_PORT">
    <xi:include href="dpram32.pb_type.xml" />
    <interconnect>
      <!-- upper -->
      <direct name="CLK_U" input="{N}_DRAM.CLK"     output="DPRAM32[1].CLK"     />
      <direct name="A_U"   input="{N}_DRAM.A[4:0]"  output="DPRAM32[1].A[4:0]"  />
      <direct name="WA_U"  input="{N}_DRAM.WA[4:0]" output="DPRAM32[1].WA[4:0]" />
      <direct name="DI2"   input="{N}_DRAM.DI2"     output="DPRAM32[1].DI"      />
      <direct name="WE_U"  input="{N}_DRAM.WE"      output="DPRAM32[1].WE"      />

      <direct name="O6"    input="DPRAM32[1].O"     output="{N}_DRAM.O6"        />
      <direct name="DO6"   input="DPRAM32[1].O"     output="{N}_DRAM.DO6_32"    />

      <!-- lower -->
      <direct name="CLK_L" input="{N}_DRAM.CLK"     output="DPRAM32[0].CLK"     />
      <direct name="A_L"   input="{N}_DRAM.A[4:0]"  output="DPRAM32[0].A[4:0]"  />
      <direct name="WA_L"  input="{N}_DRAM.WA[4:0]" output="DPRAM32[0].WA[4:0]" />
      <direct name="DI"    input="{N}_DRAM.DI1"     output="DPRAM32[0].DI"      />
      <direct name="WE_L"  input="{N}_DRAM.WE"      output="DPRAM32[0].WE"      />

      <direct name="O5"    input="DPRAM32[0].O"     output="{N}_DRAM.O5"        />
      <direct name="DO5"   input="DPRAM32[0].O"     output="{N}_DRAM.DO5_32"    />
    </interconnect>
    <metadata>
      <meta name="fasm_prefix">b{N}LUT</meta>
    </metadata>
  </mode>
  <metadata>
    <meta name="type">block</meta>
    <meta name="subtype">ignore</meta>
  </metadata>
</pb_type>
```

I got the following output;
```
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.DLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.DLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.DLUT.INIT[31:0]=32'b10010110101001011001011010100101

CLBLM_L_X20Y119.SLICEM_X0.dCLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.dCLUT.SMALL

CLBLM_L_X20Y119.SLICEM_X0.bCLUT.RAM
CLBLM_L_X20Y119.SLICEM_X0.bCLUT.SMALL
CLBLM_L_X20Y119.SLICEM_X0.bCLUT.INIT[31:0]=32'b10010110101001011001011010100101

CLBLM_L_X20Y119.SLICEM_X0.dBLUT.BLUT.INIT[31:0]=32'b11111111111111110000000000000000
CLBLM_L_X20Y119.SLICEM_X0.dBLUT.BLUT.INIT[63:32]=32'b00000000000000010000000000000001

CLBLM_L_X20Y119.SLICEM_X0.dALUT.ALUT.INIT[31:0]=32'b00001111000011110000111100001111
CLBLM_L_X20Y119.SLICEM_X0.dALUT.ALUT.INIT[63:32]=32'b00000000000000001111111111111111
```

I can't see how `dCLUT` could have `RAM` / `SMALL` as they are not under `<mode name="LUT">`. I also can't see how the lut could be in both `<mode name="LUT">` and `<mode name="32_DUAL_PORT">` modes at the same time.

So it does seems to be related to how the modes are being processed....

**Comment by mithro** [comment]
Just FYI - this still happens with the latest vpr that acomodi merged;
```
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.dCLUT.RAM not found from line 'CLBLM_L_X20Y110.SLICEM_X0.dCLUT.RAM'                                        
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.dCLUT.SMALL not found from line 'CLBLM_L_X20Y110.SLICEM_X0.dCLUT.SMALL'                                                                            
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.RAM not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.RAM'                                                                                
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.SMALL not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.SMALL'                                                                            
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.INIT not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.INIT[31:0] = 32'b10010110101001011001011010100101'                                 
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.INIT not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.INIT[31:0] = 32'b10010110101001011001011010100101'                                 
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.INIT not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.INIT[31:0] = 32'b10010110101001011001011010100101' 
```

**Comment by litghost** [comment]
> Just FYI - this still happens with the latest vpr that acomodi merged;

acomodi introduced a regression when he merged with upstream that was causing the ice40 fasm to fail.  The FASM issue present in this PR may be independent (but both related to FASM prefix handling).

Anyways, I recommend merging with symbiflow master and vtr master+wip and proceed debugging from there.

**Comment by mithro** [comment]
Same behaviour using https://github.com/SymbiFlow/symbiflow-arch-defs/blob/master/CMakeLists.txt#L45

**Comment by mithro** [comment]
@acomodi - Could you take a look at this?

The issue is most certainly in this file ->
https://github.com/SymbiFlow/vtr-verilog-to-routing/blob/master%2Bwip/utils/fasm/src/fasm.cpp

I'm reasonably sure it is in the `FasmWriterVisitor::build_clb_prefix` function;

https://github.com/SymbiFlow/vtr-verilog-to-routing/blob/2e7e9ef9306e2f50e41be88f07f0cdbc5fa9f8e2/utils/fasm/src/fasm.cpp#L139-L181

File: xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml [hunk 4]
```diff
@@ -14,50 +14,20 @@
...
+      <direct name="O5" input="{N}LUT.O5" output="{N}_DRAM.O5"/>
+      <direct name="O6" input="{N}LUT.O6" output="{N}_DRAM.O6"/>
     </interconnect>
     <metadata>
-      <meta name="fasm_type">SPLIT_LUT</meta>
-      <meta name="fasm_lut">
-        INIT[31:0] = {N}5LUT[0]
-        INIT[63:32] = {N}5LUT[1]
-      </meta>
+      <meta name="fasm_prefix">d{N}LUT</meta>
```
Comment (acomodi) [inline xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml]: @mithro is the `d` previous to {N}LUT necessary? Because it creates wrong `.fasm` output. I have run some `dram` tests after removing `d, b and c` from the `fasm_prefixes` and the issue seems to be solved. I am also running the `all_xc7` tests to see if the issue is present somewhere else, and the `dram_1_32m` correctly completes.

Comment (mithro) [reply xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml]: @acomodi - The issue I'm seeing is that we are getting **both** the `dCLUT` and `bCLUT` output for the **same** BEL -- this shouldn't be possible as the `dCLUT` and the `bCLUT` tags are on the different modes of the same block.
```
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.dCLUT.RAM not found from line 'CLBLM_L_X20Y110.SLICEM_X0.dCLUT.RAM'                                        
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.dCLUT.SMALL not found from line 'CLBLM_L_X20Y110.SLICEM_X0.dCLUT.SMALL'                                                                            
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.RAM not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.RAM'                                                                                
Segment DB CLBLM_L, key CLBLM_L.SLICEM_X0.bCLUT.SMALL not found from line 'CLBLM_L_X20Y110.SLICEM_X0.bCLUT.SMALL' 
```

Comment (mithro) [reply xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml]: The actual FASM tags which are output is not important, the important fact is that it seems to be producing output for an impossible case.

Comment (acomodi) [reply xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml]: Ok, so the addition of `d, c, b` was just for debugging purposes. And the issue here is that a same bel is expressed in two different modes, which, as you say, is an impossible case. I will double-check whether the `.net` file produced after packing is correct, to be sure that the issue is indeed in `fasm.cpp`.

Comment (mithro) [reply xc7/primitives/slicem/Ndram/ntemplate.N_dram.pb_type.xml]: @acomodi Correct!

**Comment by acomodi** [comment]
@mithro rebasing with latest master should make this PR go green (thanks to https://github.com/SymbiFlow/symbiflow-arch-defs/pull/811 and https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/69). Also DCO is failing, probably some commits haven't been signed off.

**Comment by acomodi** [comment]
@mithro CI went green, I believe it can be merged now
